doc/man/gnunet-search.1
.TH GNUNET-SEARCH "1" "22 Aug 2004" "GNUnet"
.SH NAME
gnunet-search \- a command line interface to search for content on GNUnet 
.SH SYNOPSIS
.B gnunet\-search
[\fIOPTIONS\fR] KEYWORD [AND KEYWORD]*
.SH DESCRIPTION
.PP
Search for content on GNUnet. The keywords are case-sensitive.  gnunet\-search can be used both for a search in the global namespace as well as for searching a private subspace.  
.TP
\fB\-a \fILEVEL\fR, \fB\-\-anonymity=LEVEL\fR
set additional receiver anonymity enforcing. If unset, value specified in config file is used. Default is 0.
.TP
\fB\-c \fIFILENAME\fR, \fB\-\-config=FILENAME\fR
use config file (defaults: ~/.gnunet/gnunet.conf)
.TP
\fB\-h\fR, \fB\-\-help\fR
print help page
.TP
\fB\-L \fILOGLEVEL\fR, \fB\-\-loglevel=LOGLEVEL\fR
change the loglevel. Possible values for LOGLEVEL are NOTHING, FATAL, ERROR, FAILURE, WARNING, MESSAGE, INFO, DEBUG, CRON and EVERYTHING.
.TP
\fB\-m\fR \fIMAXRESULTS\fR
exit after finding MAXRESULTS results.
.TP
\fB\-n\fR \fINAMESPACE\fR
Searches the given NAMESPACE.  The NAMESPACE can either be specified using the unique ENC identifier or using a nickname from the output of gnunet\-pseudonym.  The keyword argument specified is used as the identifier in the namespace.  Only one keyword may be specified together with the \-n option.
.TP
\fB\-o PREFIX, \fB\-\-output=PREFIX\fR
Writes the encountered (unencrypted) RBlocks or SBlocks to files with name PREFIX.XXX, where XXX is a number. This is useful to keep search results around.
.TP
\fB\-t\fR \fITIMEOUT\fR
wait TIMEOUT seconds before aborting the search.
.TP
\fB\-u\fR, \fB\-\-uri\fR
run gnunet\-search with an URI instead of a keyword.  The URI can have the format for a namespace search or for a keyword search.  For a namespace search, the format is gnunet://afs/NAMESPACE/IDENTIFIER.  For a keyword search use gnunet://afs/KEYWORD[+KEYWORD]*.
.TP
\fB\-v\fR, \fB\-\-version\fR
print the version number
.SH NOTES
The -t option specifies that the query should timeout after approximately TIMEOUT seconds. A value of zero is interpreted as never. The default value can be changed via the SEARCHTIMEOUT variable in the GNUnet configuration file. If multiple words are passed as keywords that are not separated by an AND, gnunet-search will concatenate them to one bigger keyword. Thus

# gnunet\-search Das Kapital

and

# gnunet\-search "Das Kapital"

are identical. You can use AND to separate keywords. In that case, gnunet-search will only display results that match all the keywords. gnunet-search cannot do multiple independent queries ("OR"), you must use multiple processes for that. Mind that each GNUnet-server (gnunetd) has a build-in limit on the number of clients (like gnunet-search or gnunet-download) that can connect simultaniously (you can change it in src/include/config.h).

Search results are printed by gnunet-search like this:

gnunet-download -o "COPYING" gnunet://afs/N8RCF3TETLRU9CV1PAS7M2H9QDB36AE3.K9JO8IP7KTNFO23S3VB4TFUKLD7SO5AS.0466DC92.17992
=> The GNU Public License <= (mimetype: text/plain)

The first line contains the command to run to download the file. The suggested filename in the example is COPYING. The GNUnet URI consists of the key and query hash of the file, the CRC32 (GNUnet style) and finally the size of the file. The second line contains the description of the file, here "The GNU Public License" and the mime\-type (see the options for gnunet-insert on how to change these).

.SH SETTING ANONYMITY LEVEL

The \fB-a\fR option can be used to specify additional anonymity constraints. If set to 0, GNUnet will try to download the file as fast as possible without any additional slowdown by the anonymity code. Note that you will still have a fair degree of anonymity depending on the current network load and the power of the adversary. The download is still unlikely to be terribly fast since the sender may have requested sender\-anonymity and since in addition to that, GNUnet will still do the anonymous routing.

This option can be used to limit requests further than that. In particular, you can require GNUnet to receive certain amounts of traffic from other peers before sending your queries. This way, you can gain very high levels of anonymity \- at the expense of much more traffic and much higher latency. So set it only if you really believe you need it. 

The definition of ANONYMITY-RECEIVE is the following: If the value v is < 1000, it means that if GNUnet routes n bytes of messages from foreign peers, it may originate n/v bytes of queries in the same time-period. The time\-period is twice the average delay that GNUnet deferrs forwarded queries. If the value v is >= 1000, it means that if GNUnet routes n bytes of QUERIES from at least (v % 1000) peers, it may originate n/v/1000 bytes of queries in the same time-period.

The default is 0 and this should be fine for most users. Also notice that if you choose values above 1000, you may end up having no throughput at all, especially if many of your fellow GNUnet\-peers do the same.

.SH FILES
.TP
~/.gnunet/gnunet.conf
GNUnet configuration file; specifies the default value for the timeout
.SH "REPORTING BUGS"
Report bugs by using mantis <http://www.ovmj.org/~mantis/> or by sending electronic mail to <gnunet-developers@gnu.org>
.SH "SEE ALSO"
\fBgnunet\-gtk\fP(1), \fBgnunet\-insert\fP(1), \fBgnunet\-download\fP(1), \fBgnunet\-pseudonym\fP(1), \fBgnunet.conf\fP(5), \fBgnunetd\fP(1)


ChangeLog
Mon Feb 28 00:33:03 EST 2005
	Releasing GNUnet 0.6.6b.

Sun Feb 27 19:27:44 EST 2005
	Fixing mantis 839.

Sun Feb 27 18:44:11 EST 2005
	Fixing mantis 833/834.

Sun Feb 27 18:15:02 EST 2005
	Fixing serious problem with LE in gnunet-insert.
	Fixing mantis 828.

Thu Feb 17 01:14:30 EST 2005
	Various memory leaks for Win32 fixed.
	Added support for GNUNET(D)_CONFIG environment variable.

Wed Feb  9 13:33:38 EST 2005
        Fixed problems with the logrotate code.  Releasing GNUnet 0.6.6a.

Sun Jan 11 19:58:13 CET 2005
	Releasing GNUnet 0.6.6

Sun Jan 11 19:24:27 CET 2005
        gnunet-gtk now resumes aborted downloads automatically.

Sun Nov 14 16:49:02 EST 2004
	Debugged collections, seems to work (not that it looks pretty,
	but it basically works, the rest is UI work).  Fixed deadlock
	in gnunet-gtk logging.  Releasing GNUnet 0.6.5.

Thu Nov  4 09:04:48 EST 2004
	Added (highly experimental) support for collections.

Sat Sep 25 22:29:48 EST 2004
	Releasing GNUnet 0.6.4a.

Sun Sep 19 23:39:04 EST 2004
	Various code cleanups and minor bugfixes.

Sun Aug 29 15:11:28 IST 2004
	Releasing GNUnet 0.6.4.

Tue Aug 24 20:59:14 IST 2004
	Significant enhancements to namespaces.  A namespace can
	now be annotated with meta-data at the time that it is
	created.
	
Sat Aug 21 01:00:06 IST 2004
	Adding GNU gettext support to configure.
	
Thu Aug 19 01:26:55 IST 2004
	Migrating gnunet-gtk to GTK 2.0 (tested with 2.4).

Mon Aug 16 00:30:59 IST 2004
	Removed gcry.  Added dependency on GNU MP library (libgmp)
	for kblocks.

Tue Aug  3 17:57:43 IST 2004
	Fixed bugs in gnunet-update.  Fixed segfault in AFS startup.
	Fixed warnings in gnunet-peer-info/gnunetd.  Fixed packaging
	error (not all of util/gcry included in distribution).
	Releasing GNUnet 0.6.3a.

Mon Aug  2 20:06:07 IST 2004
	Releasing GNUnet 0.6.3.

Sun Aug  1 01:31:19 IST 2004
	Changed most places from HexName to the new, shorter EncName.
	Pushed URIs for namespaces and keyword searches into more
	widespread use.  Replaced all occurences of sprintf with
	SNPRINTF (to obtain extra bounds check).
	
Sun Jul 11 13:54:01 IST 2004
	Moved testcases for libgnunet_util to util/ directory.
	Made DB testcases in afs/module/ run for each available
	database and not just the one configured in /etc/gnunet.conf.

Fri Jun 25 21:05:59 IST 2004
	Releasing GNUnet 0.6.2b (resolves various minor bugs).

Wed May  5 17:34:39 EST 2004
	Releasing GNUnet 0.6.2a (resolves libgcrypt and packageing
	issues, improved gtk-statistics, other minor bugfixes).
	
Fri Apr 30 21:36:23 EST 2004
	Fixed synchronization problem in cron (only relevant for
	parallel downloading).  Releasing GNUnet 0.6.2.

Mon Apr 26 21:13:45 EST 2004
	Added support to gnunet-check to convert file index database
	to new design.  Removed support for pre 0.6.1b bloomfilter
	conversion (would have been too messy to support both).
	Added old-version check to gnunetd startup to make updates
	more systematic in the future (not perfect, though).

Mon Apr 26 19:27:29 EST 2004
	Recursive insertion of directories with gnunet-gtk 
	now works including the progress bar.

Sun Apr 25 18:24:55 EST 2004
	Global changes to fix bug #698.  The fix is still 
	incomplete with respect to pthread.h specific structs.
	
Sun Apr 25 15:20:29 EST 2004
	Fixed compile error in IPv6 transports.

Fri Apr 23 23:38:01 EST 2004
	Added recursive insertion of directories to gnunet-gtk.
	Progress-bar does not work yet, also libextractor is
	always used (no option to turn it off).
	
Fri Apr 23 21:49:18 EST 2004
	Removed need to specify _protocol in configuration file
	(shorter is better).

Wed Apr 21 23:32:36 EST 2004
	Added code to bound priority used by clients by twice the
	average priority of requests from other peers that are 
	routed at the moment.  That prevents clients from exposing
	themselves with excessively high priorities (and also 
	avoids giving too much credit/trust needlessly).

Tue Apr 20 22:27:19 EST 2004
	Added possibility of using a link instead of making a copy
	if gnunetd and the insert-client both run on the same machine.

Tue Apr 20 00:00:58 EST 2004
	Added network transparency for indexing files (by copying
	them to a GNUnet(d) specific directory).  This breaks
	unindexing (for files indexed with previous versions) and
	forces the user to make a copy of the file.  In the near
	future we should resolve this problem by allowing the use
	of a link.

Sun Apr 18 17:24:34 EST 2004
	Adding recursive, parallel download of directories to
	gnunet-download

Sun Apr 18 01:57:32 EST 2004
	gnunet-transport-check can now be used to check
	external connectivity (with other peers).
	
Sat Apr 17 17:46:19 EST 2004
	Reduced memory consuption for AFS insertion and
	download.
	
Sat Apr 10 21:17:46 EST 2004
	Implemented statistics-plot in gnunet-gtk based on
	gnome-system-monitor code.

Fri Apr  9 22:01:51 EST 2004
	Added support for libextractor > 0.2.7.
	
Fri Apr  9  0:29:20 EST 2004
	Completed --help conversion.

Wed Mar 31 15:40:51 EST 2004
	Releasing 0.6.1d.

Tue Mar 30 22:42:13 EST 2004
	Integrated gnunet-pseudonym-create, -delete and -list into
	gnunet-pseudonym as well as gnunet-directory-list, -emptydb
	and -print into gnunet-directory.

Tue Mar  2 10:46:01 EST 2004
	Fixing bug preventing download for indexed files <= 1k.

Sat Feb 28 03:43:34 EST 2004
	Releasing 0.6.1c.

Sat Feb 21 06:16:47 EST 2004
	Modified requestmanager to improve performance at the end of the 
	download.

Thu Feb 19 00:48:33 EST 2004
	Fixed integer underflow possibly causing slow download performance.
	
Sun Feb 15 15:17:05 EST 2004
	Fixed weak key failures with libgcrypt.

Sun Feb 15 13:14:44 EST 2004
	New routing code (untested) commited.

Sun Feb 15 11:58:10 EST 2004
	Fixing problem with 64-bit size_t on alpha (#655).

Sat Jan 31 04:32:33 EST 2004
	Releasing 0.6.1b.

Fri Jan 23 00:00:09 EST 2004
	Fixed indexed content being overwritten by migrated content
	
Thu Jan 22 19:27:33 EST 2004
	Fixed integer-overflow problems in statuscalls.

src/applications/afs/tools/gnunet-search.c
/*
     This file is part of GNUnet.
     (C) 2001, 2002, 2003, 2004 Christian Grothoff (and other contributing authors)

     GNUnet is free software; you can redistribute it and/or modify
     it under the terms of the GNU General Public License as published
     by the Free Software Foundation; either version 2, or (at your
     option) any later version.

     GNUnet is distributed in the hope that it will be useful, but
     WITHOUT ANY WARRANTY; without even the implied warranty of
     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
     General Public License for more details.

     You should have received a copy of the GNU General Public License
     along with GNUnet; see the file COPYING.  If not, write to the
     Free Software Foundation, Inc., 59 Temple Place - Suite 330,
     Boston, MA 02111-1307, USA.
*/

/**
 * @file applications/afs/tools/gnunet-search.c 
 * @brief Main function to search for files on GNUnet.
 * @author Christian Grothoff
 *
 * Todo: namespace search should be more like the normal
 * search (add cron for repeated queries, allow multiple results).
 * But this requires some changes in esed2-lib first!
 *
 * OTOH, it would also probably be good if we had a convenience
 * method for printing RBlocks in esed2-lib like what we have
 * for SBlocks.
 **/

#include "gnunet_afs_esed2.h"
#include "platform.h"

typedef struct {
  unsigned int resultCount;
  unsigned int max;
} SearchClosure;

/**
 * Handle the search result.
 **/
static void handleNormalResult(RootNode * rootNode,
			       SearchClosure * sc) {
  char * fstring;
  char * fname;
  char * prefix;  

  /* write rblock to file? */
  prefix = getConfigurationString("GNUNET-SEARCH",
  				  "OUTPUT_PREFIX");
  if (prefix != NULL) {
    char * outfile;

    outfile = MALLOC(strlen(prefix)+16);
    sprintf(outfile, 
	    "%s.%03d", 
	    prefix, 
	    sc->resultCount++);
    writeFile(outfile,
    	      rootNode,
	      sizeof(RootNode),
	      "600");
    FREE(outfile);
    FREE(prefix);
  }

  sc->max--;  
  fstring = fileIdentifierToString(&rootNode->header.fileIdentifier);
 
  rootNode->header.description[MAX_DESC_LEN-1] = 0;
  rootNode->header.filename[MAX_FILENAME_LEN-1] = 0;
  rootNode->header.mimetype[MAX_MIMETYPE_LEN-1] = 0;
  
  if (0 == strcmp(rootNode->header.mimetype,
		  GNUNET_DIRECTORY_MIME)) {
    fname = expandDirectoryName(rootNode->header.filename);
  } else 
    fname = STRDUP(rootNode->header.filename);
  
  printf("gnunet-download -o \"%s\" %s\n",
	 fname,
	 fstring); 
  printf("=> %s <= (mimetype: %s)\n",
	 rootNode->header.description,
	 rootNode->header.mimetype);
  FREE(fstring);
  FREE(fname);
  if (0 == sc->max)
    run_shutdown(0);
}

typedef struct {
  HashCode160 * results;
  unsigned int resultCount;
  unsigned int max;
} NSSearchClosure;

/**
 * Handle namespace result.
 */   
static void handleNamespaceResult(SBlock * sb,
				  NSSearchClosure * sqc) {
  HashCode160 curK;
  int i;
  char * prefix;

  hash(sb, sizeof(SBlock), &curK);
  for (i=0;i<sqc->resultCount;i++)
    if (equalsHashCode160(&curK,
        &sqc->results[i])) {
      LOG(LOG_DEBUG, 
	  "DEBUG: SBlock already seen\n");
      return; /* displayed already */
    }
  GROW(sqc->results,
       sqc->resultCount,
       sqc->resultCount+1);
  memcpy(&sqc->results[sqc->resultCount-1],
         &curK,
         sizeof(HashCode160));
  printSBlock(stdout,
	      sb);
  sqc->max--;  
  /* write sblock to file */
  prefix = getConfigurationString("GNUNET-SEARCH",
  				  "OUTPUT_PREFIX");
  if (prefix != NULL) {
    char * outfile;

    outfile = MALLOC(strlen(prefix)+16);
    sprintf(outfile, 
	    "%s.%03d", 
	    prefix, 
	    sqc->resultCount-1);
    writeFile(outfile,
    	      sb,
	      sizeof(SBlock),
	      "600");
    FREE(outfile);
    FREE(prefix);
  }
  if (0 == sqc->max)
    run_shutdown(0);
}

/**
 * Prints the usage information for this command if the user errs.
 * Aborts the program.
 **/
static void printhelp() {
  static Help help[] = {
    { 'a', "anonymity", "LEVEL",
      "set the desired LEVEL of receiver-anonymity" },
    HELP_CONFIG,
    HELP_HELP,
    HELP_HOSTNAME,
    HELP_LOGLEVEL,
    { 'm', "max", "LIMIT",
      "exit after receiving LIMIT results" },
    { 'n', "namespace", "HEX",
      "only search the namespace identified by HEX" },
    { 'o', "output", "PREFIX",
      "write encountered (decrypted) search results to the file PREFIX" },
    { 't', "timeout", "TIMEOUT",
      "wait TIMEOUT seconds for search results before aborting" },
    HELP_VERSION,
    HELP_END,
  };
  formatHelp("gnunet-search [OPTIONS] KEYWORD [AND KEYWORD]",
	     "Search GNUnet for files.",
	     help);
}

/**
 * Parse the options, set the timeout.
 * @param argc the number of options
 * @param argv the option list (including keywords)
 * @return SYSERR if we should exit, OK otherwise
 **/
static int parseOptions(int argc,
			char ** argv) {
  int c;  

  while (1) {
    int option_index = 0;
    static struct GNoption long_options[] = {
      LONG_DEFAULT_OPTIONS,
      { "output",    1, 0, 'o' },
      { "anonymity", 1, 0, 'a' }, 
      { "timeout",   1, 0, 't' },
      { "max",       1, 0, 'm' },
      { "namespace", 1, 0, 'n' },
      { 0,0,0,0 }
    };    
    c = GNgetopt_long(argc,
		      argv, 
		      "a:vhdc:L:H:t:o:n:m:", 
		      long_options, 
		      &option_index);    
    if (c == -1) 
      break;  /* No more flags to process */
    if (YES == parseDefaultOptions(c, GNoptarg))
      continue;
    switch(c) {
    case 'a': {
      unsigned int receivePolicy;

      if (1 != sscanf(GNoptarg, "%ud", &receivePolicy)) {
        LOG(LOG_FAILURE,
            "FAILURE: You must pass a number to the -a option.\n");
        return -1;
      }
      setConfigurationInt("AFS",
                          "ANONYMITY-RECEIVE",
                          receivePolicy);
      break;
    }
    case 'v': 
      printf("GNUnet v%s, gnunet-search v%s\n",
	     VERSION, 
	     AFS_VERSION);
      return SYSERR;
    case 'h': 
      printhelp(); 
      return SYSERR;
    case 's':
      FREENONNULL(setConfigurationString("GNUNET-SEARCH",
      					 "NAMESPACE",
					 GNoptarg));
      break;
    case 'o':
      FREENONNULL(setConfigurationString("GNUNET-SEARCH",
      					 "OUTPUT_PREFIX",
					 GNoptarg));
      break;
    case 't': {
      unsigned int timeout;
      if (1 != sscanf(GNoptarg, "%ud", &timeout)) {
	LOG(LOG_FAILURE, 
	    "You must pass a number to the -t option.\n");
	return SYSERR;
      } else {
	setConfigurationInt("AFS",
			    "SEARCHTIMEOUT",
			    timeout);
      }
      break;
    }
    case 'm': {
      unsigned int max;
      if (1 != sscanf(GNoptarg, "%ud", &max)) {
	LOG(LOG_FAILURE, 
	    "You must pass a number to the -m option.\n");
	return SYSERR;
      } else {
	setConfigurationInt("AFS",
			    "MAXRESULTS",
			    max);
	if (max == 0) 
	  return SYSERR; /* exit... */	
      }
      break;
    }
    default: 
      LOG(LOG_FAILURE,
	  "Unknown option %c. Aborting.\n"
	  "Use --help to get a list of options.\n",
	  c);
      return SYSERR;
    } /* end of parsing commandline */
  } /* while (1) */
  if (argc - GNoptind <= 0) {
    LOG(LOG_FAILURE, 
	"FAILURE: Not enough arguments. "
	"You must specify a keyword or identifier.\n");
    printhelp();
    return SYSERR;
  }
  setConfigurationStringList(&argv[GNoptind],
			     argc-GNoptind);
  return OK;
}

/**
 * Perform a normal (non-namespace) search.
 */
static void * normalSearchMain(GNUNET_TCP_SOCKET * sock) {
  SearchClosure max;
  int i;
  int keywordCount;
  char ** keyStrings;

  max.max = getConfigurationInt("AFS",
				"MAXRESULTS");
  max.resultCount = 0;
  if (max.max == 0)
    max.max = (unsigned int)-1; /* infty */
  keywordCount = getConfigurationStringList(&keyStrings);
  searchRBlock(sock,
	       keyStrings,
	       keywordCount,
	       (SearchResultCallback)&handleNormalResult,
	       &max,
	       &testShutdown,
	       NULL);
  for (i=0;i<keywordCount;i++) 
    FREE(keyStrings[i]);  
  FREE(keyStrings);
  return NULL;
}

/**
 * Perform a namespace search.
 */
static int namespaceSearchMain(GNUNET_TCP_SOCKET * sock) {
  int ret;
  NSSearchClosure sqc;
  HashCode160 namespace;
  HashCode160 identifier;
  char * nsstring;
  char * idstring;
  char ** keyStrings;
  int kc;
  int i;
	  
  nsstring = getConfigurationString("GNUNET-SEARCH", 
				    "NAMESPACE"); 
  sqc.max = getConfigurationInt("AFS",
				"MAXRESULTS");
  if (sqc.max == 0)
    sqc.max = (unsigned int)-1; /* infty */
  hex2hash((HexName*)nsstring,
	   &namespace);
  FREE(nsstring);
  
  kc = getConfigurationStringList(&keyStrings);
  ret = 1; /* '\0' terminator */
  for (i=0;i<kc;i++)
    ret += strlen(keyStrings[i]);
  ret += kc; /* spaces! */
  idstring = MALLOC(ret);
  idstring[0] = '\0';
  for (i=0;i<kc;i++) {
    strcat(idstring, keyStrings[i]);
    FREE(keyStrings[i]);
  } 
  FREE(keyStrings);
  
  if (SYSERR == tryhex2hash(idstring,
  			    &identifier)) {
     LOG(LOG_DEBUG,
         "DEBUG: namespace ID entered is not in HEX format, using hash of ASCII text (%s).\n",
	 idstring);
     hash(idstring,
	  strlen(idstring), 
	  &identifier);
  }
  FREE(idstring);

  sqc.results = NULL;
  sqc.resultCount = 0;
  ret = searchSBlock(sock,
		     &namespace,
		     &identifier,
		     &testShutdown, 
		     NULL,
		     (NSSearchResultCallback)&handleNamespaceResult,
		     &sqc);
  if (ret == SYSERR) 
    printf("Sorry, nothing found.\n");
 
  FREENONNULL(sqc.results);
  return ret;
}

/**
 * The main function to search for files on GNUnet.
 *
 * @param argc number of arguments from the command line
 * @param argv command line arguments
 * @return return value from gnunetsearch: 0: ok, -1: error
 **/   
int main(int argc,
	 char ** argv) {
  GNUNET_TCP_SOCKET * sock;
  PTHREAD_T searchThread;
  void * unused;
  PThreadMain type;
  char * ns;
 
  if (SYSERR == initUtil(argc, argv, &parseOptions))
    return 0;
  sock = getClientSocket();
  if (sock == NULL)
    errexit("FATAL: could not connect to gnunetd.\n");
  initAnonymityPolicy(NULL);
  initializeShutdownHandlers();

  /* order of cron-jobs is important, thus '- cronMILLIS' */
  addCronJob((CronJob)&run_shutdown,
	     cronSECONDS * getConfigurationInt("AFS",
					       "SEARCHTIMEOUT") - cronMILLIS,
	     0, /* no need to repeat */
	     NULL);
  startAFSPriorityTracker();
  startCron();
  
  ns = getConfigurationString("GNUNET-SEARCH", 
			      "NAMESPACE");
  if (ns != NULL) {
    FREE(ns);
    type = (PThreadMain) &namespaceSearchMain;
  } else {
    type = (PThreadMain) &normalSearchMain;
  }

  if (0 != PTHREAD_CREATE(&searchThread, 
			  type,
			  sock,
			  8 * 1024)) 
    errexit("FATAL: failed to create search thread (%s).\n",
	    strerror(errno));
  wait_for_shutdown();
  closeSocketTemporarily(sock);
  stopCron();
  stopAFSPriorityTracker();
  delCronJob((CronJob)&run_shutdown,
	     0,
	     NULL);
  PTHREAD_JOIN(&searchThread, &unused);
  doneAnonymityPolicy();
  releaseClientSocket(sock);
  doneShutdownHandlers();
  doneUtil();
  return 0;
}

/* end of gnunet-search.c */
